### PR TITLE
Generalize or restrict a few basic operators

### LINKED ISSUE #905: Missing fallback methods for vector spaces
Given a vector space, the division by a non-zero scalar can be defined in terms of the scalar product:

```julia
/(v::SomeVectorType, λ::Number) = inv(λ) * v # where * is the implemented scalar product
```

Can this fallback method be added to Base? Currently we need to implement the fallback for all new vector spaces.

### LINKED ISSUE #43903: Missing fallback methods for vector spaces
Given a vector space, the division by a non-zero scalar can be defined in terms of the scalar product:

```julia
/(v::SomeVectorType, λ::Number) = inv(λ) * v # where * is the implemented scalar product
```

Can this fallback method be added to Base? Currently we need to implement the fallback for all new vector spaces.

### PR DESCRIPTION
**Pull request by dkarrasch**
This is an attempt to close JuliaLang/LinearAlgebra.jl#905. While I was at it, I ~~restricted the fallback~~ made the most generic fallbacks simply do what the docstring says, so right and left division operators fall back to multiplication with the inverse.

```julia
\(x,y) = adjoint(adjoint(y)/adjoint(x))
```

~~to `Number`s, and replaced `adjoint` by `conj`~~ it by `inv(x)*y`. The reason is that we don't have a method for `adjoint(x::Any)`, so this fallback is somewhat pointless (edit: unless someone guesses that overloading `adjoint` is necessary). Also, the docstring of `\` doesn't state that it can be defined via `/` and `adjoint`. For types that are not subtype of `Number` or `AbstractArray`, I believe `\` simply needs to be defined explicitly. That's what I have done with `ModInt` in the `generic` tests.

### CODE REVIEW

**Comment by dkarrasch** [comment]
One potentially annoying thing with the definition `-(x, y) = x + (-1)*y` is the following error from CI:

```julia
Test Failed at /Users/julia/buildbot/worker-tabularasa/tester_macos64/build/share/julia/test/errorshow.jl:401
  Expression: occursin("MethodError: no method matching -(::Vector{Float64}, ::Rational{$(Int)})", err_str)
   Evaluated: occursin("MethodError: no method matching -(::Vector{Float64}, ::Rational{Int64})", "MethodError: no method matching +(::Vector{Float64}, ::Rational{Int64})\nFor element-wise addition, use broadcasting with dot syntax: array .+ scalar\nClosest candidates are:\n  +(::Any, ::Any, !Matched::Any, !Matched::Any...) at operators.jl:593\n  +(!Matched::Base.TwicePrecision, ::Number) at twiceprecision.jl:290\n  +(!Matched::LinearAlgebra.UniformScaling, ::Number) at ~/buildbot/worker-tabularasa/tester_macos64/build/share/julia/stdlib/v1.9/LinearAlgebra/src/uniformscaling.jl:144\n  ...")
```

So the error message becomes slightly misleading.

File: base/operators.jl [hunk 1]
```diff
@@ -525,8 +525,10 @@ julia> identity("Well, what did you expect?")
 """
 identity(x) = x
 
-+(x::Number) = x
-*(x::Number) = x
++(x) = x
+-(x) = (-1)*x
```
Comment (martinholters) [inline base/operators.jl:529]: Is `1` of the desired type here? Could `Int8(1)` reduce the risk of unwanted widening promotions? Or should the type depend on (the type of) `x` in some way? Then it is probably hard to write this generically.

Comment (dkarrasch) [reply base/operators.jl:529]:
```suggestion
-(x) = Int8(-1)*x
```

Comment (dkarrasch) [reply base/operators.jl]: Yes, I agree. If we go with this approach, then the type should be as small as possible.

**Comment by dkarrasch** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/44564#issuecomment-1083599383) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/dcab66a_vs_3f23c45/report.html).

**Comment by StefanKarpinski** [comment]
Does this need review to move forward? If so, any idea who?

**Review by stevengj — APPROVED** [review]

File: NEWS.md [hunk 2]
```diff
@@ -14,6 +14,11 @@ Language changes
 * New builtins `getglobal(::Module, ::Symbol[, order])` and `setglobal!(::Module, ::Symbol, x[, order])`
   for reading from and writing to globals. `getglobal` should now be preferred for accessing globals over
   `getfield`. ([#44137])
+* A few basic operators have been generalized to more naturally support vector space structures:
+  `+(x) = x`, unary minus falls back to scalar multiplication with -1, `-(x) = Int8(-1)*x`,
+  binary minus falls back to addition `-(x, y) = x + (-y)`, and, at the most generic level,
+  left- and right-division fall back to multiplication with the inverse from left and right,
+  resp., as stated in the docstring. ([#44564])
```
Comment (vtjnash) [inline NEWS.md:21]:
```suggestion
  respectively, as stated in the docstring. ([#44564])
```

**Comment by vtjnash** [comment]
I don't know the math sufficiently well to claim whether these are equivalent

File: base/operators.jl [hunk 3]
```diff
@@ -525,8 +525,10 @@ julia> identity("Well, what did you expect?")
 """
 identity(x) = x
 
-+(x::Number) = x
-*(x::Number) = x
++(x) = x
+-(x) = Int8(-1)*x
```
Comment (vtjnash) [inline base/operators.jl:529]: this seems possibly incorrect for many types (e.g. Unsigned), and less likely for promotion to be defined than negation

**Comment by dkarrasch** [comment]
Are the ambiguity test failures in linuxaarch64 real? If not (tests pass on all other platforms or fail for obviously unrelated reasons), then this can be merged.

**Comment by bkamins** [comment]
This PR breaks DataFrames.jl tests (fortunately tests only). Before I start fixing DataFrames.jl could you please explain me the rationale behind allowing for example:
```
julia> +(nothing)

julia> *(nothing)

julia> +("some string")
"some string"

julia> using Dates

julia> *(Date(2020, 5, 15))
2020-05-15
```

Thank you!

CC @nalimilan

**Comment by dkarrasch** [comment]
Is the failing test a new one? DataFrames.jl was not listed as failing in the nanosoldier run from March. I'm very sorry for the breakage!

The general idea of this PR is to make vector space calculus the generic default as discussed in #43903. I might have overdone it with `+` and `*`, but I felt that if we have a generic `-`, then the same should apply to the other basic operations. I'll open a draft PR in which I restrict `+` and `*` back to `Number`, and then we can choose whether we want to revert this one or go forward with the other.

**Pull request merged**